clippy_lints/src/methods/parsed_string_literals/mod.rs

```diff
@@ -0,0 +1,93 @@
+use rustc_errors::{Applicability, SubdiagMessage};
+use rustc_hir::{self as hir, AmbigArg, Expr, ExprKind, GenericArg, Node};
+use rustc_lint::LateContext;
+
+mod ip_addresses;
+mod primitive_types;
+
+use super::PARSED_STRING_LITERALS;
+
+/// Detects instances of `"literal".parse().unwrap()`:
+/// - `expr` is the whole expression
+/// - `recv` is the receiver of `parse()`
+/// - `parse_call` is the `parse()` method call
+/// - `msrv` is used for Rust version checking
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, recv: &Expr<'_>, parse_call: &Expr<'_>, msrv: Msrv) {
+    if let ExprKind::Lit(lit) = recv.kind
+        && let LitKind::Str(lit, _) = lit.node
+    {
+        let ty = cx.typeck_results().expr_ty(expr);
+        match get_type_diagnostic_name(cx, ty) {
+            _ if ty.is_primitive() => primitive_types::check(cx, expr, lit, ty, recv),
+            Some(method @ (sym::IpAddr | sym::Ipv4Addr | sym::Ipv6Addr)) => {
+                ip_addresses::check(cx, expr, lit, method, get_explicit_type(cx, expr, parse_call), msrv);
+            },
+            _ => (),
+        }
+    }
+}
+
+/// Emit the lint if the length of `sugg` is no longer than the original `expr` span, or if `force`
+/// is set.
+fn maybe_emit_lint(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    force: bool,
+    entity: &str,
+    sugg: String,
+    applicability: Applicability,
+) {
+    if force || expr.span.check_source_text(cx, |snip| snip.len() >= sugg.len()) {
+        span_lint_and_sugg(
+            cx,
+            PARSED_STRING_LITERALS,
+            expr.span,
+            format!("unnecessary runtime parsing of {entity}"),
+            "use",
+            sugg,
+            applicability,
+        );
+    }
+}
+
+/// Warn and help about an expression.
+fn warn_about(cx: &LateContext<'_>, expr: &Expr<'_>, entity: &str, help: impl Into<SubdiagMessage>) {
+    span_lint_and_help(
+        cx,
+        PARSED_STRING_LITERALS,
+        expr.span,
+        format!("unnecessary runtime parsing of {entity}"),
+        None,
+        help,
+    );
+}
+
+/// Returns `T` from the `parse::<T>(…)` call if present, or from a `let _: T =
+/// "…".parse().unwrap();` otherwise if it can be found. The `let` statement has been chosen because
+/// it is quite common to assign the result of parsing to a variable.
+fn get_explicit_type<'hir>(
+    cx: &'hir LateContext<'_>,
+    expr: &'hir Expr<'_>,
+    parse_call: &'hir Expr<'_>,
+) -> Option<hir::Ty<'hir, AmbigArg>> {
+    if let ExprKind::MethodCall(parse, _, _, _) = parse_call.kind
+        && let [GenericArg::Type(ty)] = parse.args().args
+        && matches!(ty.kind, hir::TyKind::Path(_))
+    {
+        Some(**ty)
+    } else if let Node::LetStmt(let_stmt) = cx.tcx.parent_hir_node(expr.hir_id)
+        && let Some(ty) = let_stmt.ty
+        && let Some(ty) = ty.try_as_ambig_ty()
+        && matches!(ty.kind, hir::TyKind::Path(_))
+    {
+        Some(*ty)
+    } else {
+        None
+    }
```

Comment:
Why are you looking for the parent statement here? If the type isn't on the `parse` call itself then type inference will work just fine.

Reply:
Because I hadn't noticed that `IpAddr` could get direct conversion from `[u8; 4]` (IPv4) or `[u16; 8]` (IPv6), and thought I had to go through one intermediate value. I will fix this, this will be even better then.